File: Design Patterns/Behavioral/State.ts
interface IState {
    insertCoin(): void
    selectItem(): void
}


class NoCoinState implements IState {
    constructor(private machine: VendingMachine) { }

    insertCoin() {
        console.log("Coin inserted")
        this.machine.setState(this.machine.hasCoinState)
    }

    selectItem() { console.log("Insert coin first") }
}

class HasCoinState implements IState {
    constructor(private machine: VendingMachine) { }

    insertCoin() { console.log("Coin already inserted") }

    selectItem() {
        console.log("Dispensing item")
        this.machine.setState(this.machine.noCoinState)
    }
}


class VendingMachine {
    noCoinState: IState
    hasCoinState: IState
    private currentState: IState

    constructor() {
        this.noCoinState = new NoCoinState(this)
        this.hasCoinState = new HasCoinState(this)
        this.currentState = this.noCoinState
    }

    setState(state: IState) { this.currentState = state }
    insertCoin() { this.currentState.insertCoin() }
    selectItem() { this.currentState.selectItem() }
}

const machine = new VendingMachine()

machine.selectItem() // Insert coin first
machine.insertCoin() // Coin inserted
machine.selectItem() // Dispensing item

File: Design Patterns/Behavioral/C#/State.cs
using System;

interface IState
{
    void InsertCoin();
    void SelectItem();
}

class NoCoinState : IState
{
    private readonly VendingMachine Machine;

    public NoCoinState(VendingMachine Machine)
    {
        this.Machine = Machine;
    }

    public void InsertCoin()
    {
        Console.WriteLine("Coin inserted");
        this.Machine.SetState(this.Machine.hasCoinState);
    }

    public void SelectItem()
    {
        Console.WriteLine("Insert coin first");
    }
}

class HasCoinState : IState
{
    private readonly VendingMachine Machine;

    public HasCoinState(VendingMachine Machine)
    {
        this.Machine = Machine;
    }

    public void InsertCoin()
    {
        Console.WriteLine("Coin already inserted");
    }

    public void SelectItem()
    {
        Console.WriteLine("Dispensing item");
        this.Machine.SetState(this.Machine.noCoinState);
    }
}

class VendingMachine
{
    public IState noCoinState;
    public IState hasCoinState;
    private IState currentState;

    public VendingMachine()
    {
        this.noCoinState = new NoCoinState(this);
        this.hasCoinState = new HasCoinState(this);
        this.currentState = this.noCoinState;
    }

    public void SetState(IState state) { this.currentState = state; }
    public void InsertCoin() { this.currentState.InsertCoin(); }
    public void SelectItem() { this.currentState.SelectItem(); }
}

class Program
{
    public static void Main(string[] args)
    {
        VendingMachine machine = new VendingMachine();
        machine.SelectItem();  // Output: Insert coin first
        machine.InsertCoin();  // Output: Coin inserted
        machine.SelectItem();  // Output: Dispensing item
    }
}